core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2019-2022 JetBrains s.r.o.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime
+
+import kotlin.random.Random
+import kotlin.random.nextInt
+
+internal class LocalDateProgressionIterator(private val iterator: IntIterator) : Iterator<LocalDate> {
+    override fun hasNext(): Boolean = iterator.hasNext()
+    override fun next(): LocalDate = LocalDate.fromEpochDays(iterator.next())
+}
+
+public open class LocalDateProgression
+internal constructor(internal val intProgression: IntProgression) : Iterable<LocalDate> {
+
+    internal constructor(
+        start: LocalDate,
+        endInclusive: LocalDate,
+        step: Int
+    ) : this(IntProgression.fromClosedRange(start.toEpochDays(), endInclusive.toEpochDays(), step))
```

Comment:
This implementation is incorrect for dates whose `toEpochDays()` does not fit into an `Int`. See https://kotlinlang.org/api/kotlinx-datetime/kotlinx-datetime/kotlinx.datetime/-local-date/to-epoch-days.html

Reply:
Why? Using `IntProgression` looks like just an implementation detail, not some fundamental issue. It would be better to just support all dates without relying on `toEpochDays()`.

Reply:
It looks like the bounds of `LocalDate` vary by implementation (`native` goes from year `-999_999` to `999_999`, while `jvm` goes from year `-999_999_999` to `999_999_999`) but each internally convert to epoch days in order to do arithmetic. However those internal conversions to epoch days can be broader than the `common` conversion to epoch days. I do wonder if it would be better for the `common` `toEpochDays()` to return a `Long` instead of an `Int`, so that it is not more restricted than some implementations of `LocalDate` itself. 

An alternative implementation is to switch to a `LongProgression` between `0` and `size of progression` and use `fun next() = startDate.plus(iterator.next(), DateTimeUnit.DAY)`, which would bypass the `common` `toEpochDays()` and `fromEpochDays()`. However `daysUntil` _also_ returns an `Int`, and so the size of the range can't be reliably computed either. Without the ability to compute the size of the progression, we also lose the ability to generate random values from within the progression without doing something hideous like materializing the entire progression. 

I could attempt to reimplement it without leaning on any of the primitive progressions at all, but we would still lose the ability to calculate the last element (again excepting a hideous full iteration) for steps other than 1 day and -1 day, as we can't calculate the `differenceModulo` without an `Int`/`Long` representation of `LocalDate`. 

It seems like our options are:

1. Change the return type of several `common` methods from `Int` to `Long` to ensure that the full range of `LocalDate` can be expressed as days without overflowing.
2. Accept that the use of `Int` in `common` clamps the allowable ranges tighter than some implementations of `LocalDate` itself.
3. Push everything down and implement `LocalDateProgression` on a per-platform basis.

I'm inclined to leave it clamped for now, and then unclamp it in the future if/when the `common` day methods are upgraded from `Int` to `Long`.

Reply:
> I do wonder if it would be better for the common toEpochDays() to return a Long instead of an Int, so that it is not more restricted than some implementations of LocalDate itself.

Maybe it would, but that's a separate problem. We're considering making the ranges the same, and it's possible that `toEpochDays()` will naturally have to return a `Long`: https://github.com/Kotlin/kotlinx-datetime/issues/432

> However `daysUntil` also returns an Int, and so the size of the range can't be reliably computed either.

Inside the library, we're not bound by the `public` API. If you need a `daysUntil` that returns a `Long`, you can implement it using something like

```
// common code
internal expect fun LocalDate.daysUntilLong(other: LocalDate): Long

// JVM
internal actual fun LocalDate.daysUntilLong(other: LocalDate): Long =
    this.value.until(other.value, ChronoUnit.DAYS)

// other platforms
internal actual fun LocalDate.daysUntilLong(other: LocalDate): Long =
    daysUntil(other).toLong()
```

Reply:
I added platform-specific `toEpochDaysLong()` and `fromEpochDays(epochDays: Long)` functions, and switched from an `IntProgression` to a `LongProgression`. So far, so good. The one issue that remains is that, when implementing `Collection<LocalDate>`, it expects `size` to be an `Int`. Because we have expanded `LocalDateProgression` beyond the bounds of `Int`, it is now possible to create progressions of greater size than `Int.MAX_VALUE`.

In light of this, do we still want to implement `Collection`, and if so, do we want to throw in these cases, or clamp the size to `Int`?

Reply:
I think we should clamp the size to `Int.MAX_VALUE` for consistency with Java (and, most likely, the other implementations of `Collection` we have with the same problem): https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html#size() @ilya-g, WDYT?

Reply:
It turns out we (the Kotlin library team) have already decided to use the same contract as Java here, I just didn't know that.